### ISSUE TITLE
Renv - State of the union December 2023

### ISSUE THREAD

**Issue by ddntl** [NISRA-Tech-Lab/rap-skeleton#5]
A few notes around ongoing issues and considerations for renv and the skeleton more widely.

- **updates and pending updates to the skeleton since March 2020**
A number of commits have been made to the skeleton by @brianq-nisra  and @kham1508 e.g. 93bcd5f3a335a601027cae95dbb529a1f55f6ace, 9fe112a442e2cfc34910dcd8dd2acceef329815b and a55a6dd4cb4d2dbad4d53332a627d662e48effad. There hasn't been a new 'release' since then. @kham1508 has been working on a branch to update packages, including renv on `update-renv103-and-packages`

- **ISSUE 1 - renv >= 1.0.0 and package management**
Since [1.0.0](https://github.com/rstudio/renv/releases/tag/v1.0.0), renv has moved its default package management source to a posit.co URL which seems to be blocked for staff. New projects created from scratch will default to it, but legacy projects should use the old url. **Temporary fix** modify your renv.lock file, changing the 'Repositories' 'URL' to `https://cran.rstudio.com`. **Long term fix** Currently working with IT to fix the block

- **ISSUE 2 - old rap-skeleton or RAP users can't easily benefit from the improvements to the skeleton**
As far as we're aware, all Accessibility Exemplar and RAP skeleton projects have resulted in customers either creating their own, unlinked github repos for their publications or using a shared drive. As such, there isn't an easy way for them to benefit from the commits that are made after their code 'breaks' from the rap-skeleton. **Possible solution** if a customer is using git, we could test integrating commits from our repo into their theirs.

- **ISSUE 3 renv (generally) and onboarding new users i.e. package source compilation issues**
The benefit of renv is that you can set up a project today and it will work forever. At the moment, this isn't the case for all users. There seem to be a two factors in this
  1. **Free and open source software (FOSS), independently developed dependencies and stability**
      PackageA depends on PackageB. PackageA adds new features and relies on something in PackageB, but PackageB has a bug with Windows or it has updated its package for R 4.3 but not R 4.2 (the version we use). **Solution** - there isn't much we can do here as its a case-by-case issue. Rolling back PackageA to a known working configuration is the easiest usually. Using git and having a full history of renv.lock file is important here.
  2. **Building packages from source**
      renv is powerful because it keeps a record of which version of a package you use and you can always retrieve that from the internet. CRAN (the R repository) only keeps a 'binary'[^1] (i.e. a pre-compiled, ready to rock version) of packages for the latest version of that package. Any older versions are maintained by CRAN via 'source code'[^1] (i.e. you have to build it yourself). Currently, our systems cannot always reliably build packages from source. Sometimes this is because of some technical issue with our IT, sometimes it is a bug for all users of R. **Temporary Solution** The easiest solution is to test `renv::update()` on your project i.e. update all packages to the latest version and hope for the best. **Long term solution simple** get our systems reliably building packages from source. **Long term, better solution** host and maintain a NISRA internal package repository where we can keep binarys of all the packages we use forever.

[^1]: Think of it as binary = a sponge cake you bought in the supermarket and source = buying the ingredients and baking it yourself.

**Comment by kham1508** [comment]
@ddntl The Temporary solution proposed for the Building packages from source point. Would that be our suggested first try at a fix if a package wasn't compiling? Would we not try the most recent version of that particular package first? If we just run straight to `renv::update()` we've thrown away the lockfile at the first hurdle?

**Comment by ddntl** [comment]
> @ddntl The Temporary solution proposed for the Building packages from source point. Would that be our suggested first try at a fix if a package wasn't compiling? Would we not try the most recent version of that particular package first? If we just run straight to `renv::update()` we've thrown away the lockfile at the first hurdle?

Good point, updated it to have update a few individual packages initially and next stage is update it all

**Comment by kham1508** [comment]
> > @ddntl The Temporary solution proposed for the Building packages from source point. Would that be our suggested first try at a fix if a package wasn't compiling? Would we not try the most recent version of that particular package first? If we just run straight to `renv::update()` we've thrown away the lockfile at the first hurdle?
> 
> Good point, updated it to have update a few individual packages initially and next stage is update it all

All looks fine to me to share with Erin/rest of TL
